goblint/analyzer
# Relational int domains in base

### Summary:
Added the possibility of using relational int domains in the base analysis. When adding now for example `--set ana.int.relational true --set ana.int.equations true`, a new relational domain is used, which handles relations of the form `c1 * a + c2 + b + c3 = 0`.

### Explanation:
I wrote an interface for relational int domains, similar to the IntDomTuple in the IntDomain and created two relational int domains for it:
1. SimpleEquations is a relational domain, which handles relations of the form `c1 * a + c2 + b + c3 = 0` as `(Cil.varinfo * const) * (Cil.varinfo * const) * const`. An equation is always ordered alphabetically by the variable names and c1 is always 1.
2. The ApronDomain was adjusted, such that it can be used now from the poly and the base analysis. In order to handle global, as well as local variables, the local variable names get a space `" "` added. This can hardly be noticed by the user in the printed abstract value and spaces will never be part of a variable name in C.

### Comment:
Unfortunately, I had to use IntDomain.IntDomTuple everywhere, where ValueDomain.ID was used before, because else there was a circular build error (which I could not solve otherwise).

### How to test:
`goblint --html -o ../results/results_relational_int_own/  tests/regression/01-cpa/27-interval.c  --set ana.int.relational true --set ana.int.equations true --set ana.int.aprondomain true`
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/goblint/analyzer/pull/47%23issuecomment-197842878%22%2C%20%22https%3A//github.com/goblint/analyzer/pull/47%23issuecomment-199185981%22%2C%20%22https%3A//github.com/goblint/analyzer/pull/47%23issuecomment-199210482%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/goblint/analyzer/pull/47%23issuecomment-197842878%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%27ll%20check%20this%20on%20Monday.%22%2C%20%22created_at%22%3A%20%222016-03-17T11%3A47%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/493741%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vogler%22%7D%7D%2C%20%7B%22body%22%3A%20%22Should%20I%20adjust%20the%20make.sh%20file%2C%20such%20that%20Apron%20is%20always%20built%3F%20%28This%20is%2C%20why%20the%20checks%20have%20failed.%29%22%2C%20%22created_at%22%3A%20%222016-03-21T09%3A13%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8916408%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/isabelmax%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%27ll%20try%20to%20patch%20it%20later%2C%20so%20that%20Apron%20isn%27t%20required%20by%20default.%22%2C%20%22created_at%22%3A%20%222016-03-21T10%3A11%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/493741%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vogler%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/goblint/analyzer/pull/47#issuecomment-197842878'>General Comment</a></b>
- <a href='https://github.com/vogler'><img border=0 src='https://avatars.githubusercontent.com/u/493741?v=3' height=16 width=16'></a> I'll check this on Monday.
- <a href='https://github.com/isabelmax'><img border=0 src='https://avatars.githubusercontent.com/u/8916408?v=3' height=16 width=16'></a> Should I adjust the make.sh file, such that Apron is always built? (This is, why the checks have failed.)
- <a href='https://github.com/vogler'><img border=0 src='https://avatars.githubusercontent.com/u/493741?v=3' height=16 width=16'></a> I'll try to patch it later, so that Apron isn't required by default.


<a href='https://www.codereviewhub.com/goblint/analyzer/pull/47?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/goblint/analyzer/pull/47?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/goblint/analyzer/pull/47'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/goblint/analyzer/47)

***
***
***
